### PR TITLE
Feature: setup project

### PR DESCRIPTION
**Pull request by Zorin95670**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,62 @@
+{
+  "name": "leto-modelizer-api",
+  "version": "0.1.0",
+  "description": "Backend for leto-modelizer to manage libraries and access rights management.",
```
Comment (k-y0u) [inline package.json:4]: remove `management`

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import http from 'http';
+import express from 'express';
+import { ParseServer } from 'parse-server';
+import { loadConfiguration } from './services/ConfigurationService.js';
```
Comment (k-y0u) [inline src/index.js:4]: `src/services/ConfigurationService.js`

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+const configuration = loadConfiguration();
+const app = express();
+const server = new ParseServer(configuration.parseServer);
+const httpServer = http.createServer(app);
+
+await server.start();
+app.use(configuration.mountPath, server.app);
+
+app.get('/', (req, res) => {
+  res.status(200).send('I dream of being a website.  Please star the parse-server repo on GitHub!');
```
Comment (k-y0u) [inline src/index.js:15]: what is this message ? XD

File: src/models/Configuration.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
+/**
+ * All api configuration key and value.
```
Comment (k-y0u) [inline src/models/Configuration.js:2]: keys and values

File: tests/unit/models/Configuration.spec.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
+import Configuration from 'src/models/Configuration.js';
+
+describe('Test class: Configuration', () => {
+  describe('Test constructor', () => {
+    it('Checks variable instantiation', () => {
```
Comment (k-y0u) [inline tests/unit/models/Configuration.spec.js:5]: variables

File: .eslintrc.cjs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    sourceType: 'module'
+  },
+
+  env: {
+    jest: true,
+  },
+
+  extends: [
+    'airbnb-base',
+    "plugin:cypress/recommended"
```
Comment (justinehell) [inline .eslintrc.cjs]: why double quote instead of single one ?

File: src/models/Configuration.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+ */
+class Configuration {
+  /**
+   * Default constructor.
+   * @param {object} [props] - Object that contains all properties to set.
+   * @param {string} [props.mode] - The name of your .env file and running mode.
+   * @param {string} [props.mountPath] - That specifies the mount path where the Parse middleware
+   * should be deployed in your Node.js application.
+   * @param {number} [props.port] - The network port number on which a server process listens
+   * for incoming connections from clients.
```
Comment (justinehell) [inline src/models/Configuration.js:12]:
```suggestion
   * for incoming client connections.
```

File: src/models/Configuration.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
+/**
+ * All api configuration key and value.
+ */
+class Configuration {
+  /**
+   * Default constructor.
+   * @param {object} [props] - Object that contains all properties to set.
+   * @param {string} [props.mode] - The name of your .env file and running mode.
+   * @param {string} [props.mountPath] - That specifies the mount path where the Parse middleware
```
Comment (justinehell) [inline src/models/Configuration.js:9]: Missing .

File: src/models/Configuration.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+   * should be deployed in your Node.js application.
+   * @param {number} [props.port] - The network port number on which a server process listens
+   * for incoming connections from clients.
+   * @param {object} [props.parseServer] - Parse server configuration object.
+   * @param {string} [props.parseServer.databaseURI] - Represents a connection string used to
+   * connect to a database.
+   * @param {string} [props.parseServer.appId] - A unique identifier for your application,
+   * used to distinguish it when interacting with the Parse backend
```
Comment (justinehell) [inline src/models/Configuration.js:17]: missing .

File: src/models/Configuration.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    this.parseServer.databaseURI = props.parseServer?.databaseURI || null;
+    /**
+     * A unique identifier for your application, used to distinguish it when interacting with
+     * the Parse backend.
+     * @type {string}
+     */
+    this.parseServer.appId = props.parseServer?.appId || 'leto-modelizer-api';
+    /**
+     * A special access key that grants unrestricted access to the Parse backend, allowing
+     * administrative-level operations and data management
```
Comment (justinehell) [inline src/models/Configuration.js:70]: Missing .

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+- Commit the modifications with the label `Release version X.Y.Z`.
+- Create a pull request on github for this branch into `main`.
+- Once the pull request validated and merged, tag the `main` branch with `vX.Y.Z`.
+- After the tag is pushed, make the release on the tag in GitHub.
+
+### Git: Default branch
+
+The default branch is main. Direct commit on it is forbidden. The only way to update the application is through pull request.
+
+Release tag are only done on the `main` branch.
```
Comment (justinehell) [inline CONTRIBUTING.md:115]: Release tags are ...

**Review by justinehell — CHANGES_REQUESTED** [review]

File: README.md [hunk 12]
```diff
@@ -1,2 +1,83 @@
...
+cd leto-modelizer-api
+```
+
+3. Set up Environment Variables: Create a `prod.env` file in the root of the project directory to store your
+   configuration settings. Add the following variables to the .env file:
+
+```makefile
+APP_ID=my_app_id
+MASTER_KEY=my_super_secret_master_key
+SERVER_URL=http://localhost:1337/api
```
Comment (justinehell) [inline README.md:40]: Maybe add here at least **all the required** variables
At first, I thought these three were enough.

File: README.md [hunk 13]
```diff
@@ -1,2 +1,83 @@
...
+
+### Installation
+
+To get started with __leto-modelizer-api__ using Docker, follow these steps:
+
+1. Install Docker: If you don't have Docker installed on your system, you can download and install it from the official
+   website for your operating system:
+    * [Docker Desktop for Windows]()
+    * [Docker Desktop for macOS]()
+    * [Docker Engine for Linux]()
```
Comment (justinehell) [inline README.md:24]: add links

File: README.md [hunk 14]
```diff
@@ -1,2 +1,83 @@
...
+```shell
+docker build --build-arg NODE_ENV=prod -t leto-modelizer-api .
+ 
+docker-compose up -d
+```
+
+This command will build leto-modelizer-api set up the server with the specified configuration.
+
+Verify Installation: Once the containers are up and running, you can access your servver at the defined
+SERVER_URL (http://localhost:1337/api in this example).
```
Comment (justinehell) [inline README.md:58]: image

**Review by justinehell — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (Zorin95670) [reply package.json:4]: done

File: src/index.js (hunk 2 above)
Comment (Zorin95670) [reply src/index.js:4]: Nope

File: README.md [hunk 15]
```diff
@@ -1,2 +1,83 @@
...
+To get started with __leto-modelizer-api__ using Docker, follow these steps:
+
+1. Install Docker: If you don't have Docker installed on your system, you can download and install it from the official
+   website for your operating system:
+    * [Docker Desktop for Windows]()
+    * [Docker Desktop for macOS]()
+    * [Docker Engine for Linux]()
+
+
+2. Clone the repository: Open your terminal or command prompt and clone the project repository to your local machine:
```
Comment (k-y0u) [inline README.md:27]: to your machine

File: README.md (hunk 12 above)
Comment (k-y0u) [inline README.md:40]: `NODE_ENV` missing but indicate as required below.

File: src/index.js [hunk 16]
```diff
@@ -0,0 +1,23 @@
+import http from 'http';
+import express from 'express';
+import { ParseServer } from 'parse-server';
+import { loadConfiguration } from './services/ConfigurationService.js';
+
+const configuration = loadConfiguration();
+const app = express();
+const server = new ParseServer(configuration.parseServer);
+const httpServer = http.createServer(app);
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/index.js:9]: done

File: README.md [hunk 17]
```diff
@@ -1,2 +1,83 @@
...
+
+```bash
+git clone git+https://github.com/ditrit/leto-modelizer-api.git
+cd leto-modelizer-api
+```
+
+3. Set up Environment Variables: Create a `prod.env` file in the root of the project directory to store your
```
Comment (k-y0u) [inline README.md:34]: Couldn't it be `dev.env`, `staging.env` or else ?

File: src/index.js (hunk 3 above)
Comment (Zorin95670) [reply src/index.js:15]: done

File: src/models/Configuration.js (hunk 4 above)
Comment (Zorin95670) [reply src/models/Configuration.js:2]: done

File: tests/unit/models/Configuration.spec.js (hunk 5 above)
Comment (Zorin95670) [reply tests/unit/models/Configuration.spec.js:5]: done

File: README.md [hunk 18]
```diff
@@ -1,2 +1,83 @@
 # leto-modelizer-api
-Backend for leto-modelizer to manage libraries and access rights management
+
+Backend for leto-modelizer to manage libraries and access rights management.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply README.md:3]: done

File: .eslintrc.cjs (hunk 6 above)
Comment (Zorin95670) [reply .eslintrc.cjs]: done

File: src/models/Configuration.js (hunk 7 above)
Comment (Zorin95670) [reply src/models/Configuration.js:12]: done

File: src/models/Configuration.js (hunk 8 above)
Comment (Zorin95670) [reply src/models/Configuration.js:9]: done

File: src/models/Configuration.js (hunk 10 above)
Comment (Zorin95670) [reply src/models/Configuration.js:70]: done

File: CONTRIBUTING.md (hunk 11 above)
Comment (Zorin95670) [reply CONTRIBUTING.md:115]: done

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+Run all the unit tests.
+
+### test:coverage
+
+Run all the unit tests and generate coverage report of the unit tests for sonar.
+
+### test:e2e
+
+To run all the e2e tests, you need to run the application in dev mode.
+## Development
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply CONTRIBUTING.md:68]: done

File: README.md [hunk 20]
```diff
@@ -1,2 +1,83 @@
...
+   website for your operating system:
+    * [Docker Desktop for Windows]()
+    * [Docker Desktop for macOS]()
+    * [Docker Engine for Linux]()
+
+
+2. Clone the repository: Open your terminal or command prompt and clone the project repository to your local machine:
+
+```bash
+git clone git+https://github.com/ditrit/leto-modelizer-api.git
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply README.md:30]: done

File: README.md [hunk 21]
```diff
@@ -1,2 +1,83 @@
...
+
+```shell
+docker-compose down
+```
+
+Now your project is up and running using Docker, providing a convenient and isolated environment for your backend.
+
+### Configuration
+
+| Variable     | Optional ?                                | Example                                               | Description                                                                                                                              |
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply README.md:71]: done

File: README.md [hunk 22]
```diff
@@ -1,2 +1,83 @@
...
+
+```shell
+docker build --build-arg NODE_ENV=prod -t leto-modelizer-api .
+ 
+docker-compose up -d
+```
+
+This command will build leto-modelizer-api set up the server with the specified configuration.
+
+Verify Installation: Once the containers are up and running, you can access your servver at the defined
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply README.md:57]: done

File: README.md (hunk 12 above)
Comment (k-y0u) [inline README.md:40]: `DATABASE_URI` missing but indicate as required below.

Comment (Zorin95670) [reply README.md:40]: done

File: README.md (hunk 13 above)
Comment (Zorin95670) [reply README.md:24]: done

File: README.md (hunk 14 above)
Comment (Zorin95670) [reply README.md:58]: Normal you must provide authentication :)

File: README.md (hunk 15 above)
Comment (Zorin95670) [reply README.md:27]: done

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (Zorin95670) [reply README.md:40]: Not done, set all variable to make an example

Comment (Zorin95670) [reply README.md:40]: done

File: README.md (hunk 17 above)
Comment (Zorin95670) [reply README.md:34]: you could name it as you want, but you have to use same name in `NODE_ENV`, as explain in documentation :)

File: README.md (hunk 12 above)
Comment (Zorin95670) [reply README.md:40]: done

File: src/index.js (hunk 2 above)
Comment (k-y0u) [reply src/index.js:4]: what's the link ?

**Review by justinehell — DISMISSED** [review]

File: README.md [hunk 23]
```diff
@@ -1,2 +1,104 @@
...
+Open your terminal or command prompt and clone the project repository to your machine:
+
+```bash
+git clone git@github.com:ditrit/leto-modelizer-api.git
+cd leto-modelizer-api
+```
+
+### Run server
+
+Build and Run with Docker Compose: Run the following command to build the Docker containers and start the server:
```
Comment (k-y0u) [inline README.md:43]: step one

File: README.md [hunk 24]
```diff
@@ -1,2 +1,104 @@
...
+
+```shell
+docker build --build-arg NODE_ENV=dev -t leto-modelizer-api .
+ 
+docker-compose up -d
+```
+
+This command will build leto-modelizer-api set up the server with the specified configuration.
+
+Verify Installation: Once the containers are up and running, you can access your server at the defined
```
Comment (k-y0u) [inline README.md:53]: step two

File: README.md [hunk 25]
```diff
@@ -1,2 +1,104 @@
...
+SERVER_URL (http://localhost:1337/api/health in this example).
+You can now start using the Parse functionalities in your project.
+
+Clean Up: To stop and remove the Docker containers when you're done using the project, run:
```
Comment (k-y0u) [inline README.md:57]: step three

File: README.md [hunk 26]
```diff
@@ -1,2 +1,104 @@
...
+PORT=1337
+NODE_ENV=prod
+```
+
+See Configuration section for more details.
+
+Then build your docker image.
+
+```shell
+docker build --build-arg -t leto-modelizer-api .
```
Comment (k-y0u) [inline README.md:97]: forgot `NODE_ENV=[env_name]`

File: README.md [hunk 27]
```diff
@@ -1,2 +1,104 @@
...
+| SERVER_URL   | No, default: `http://localhost:1337/api` | http://localhost:1337/api                             | That defines the URL where the server is hosted, enabling clients to connect and interact with the Parse backend.                        |
+| PARSE_MOUNT  | No, default: `/api`                      | /api                                                  | That specifies the mount path where the Parse middleware should be deployed in your Node.js application.                                 |
+| PORT         | No, default: `1337`                      | 1337                                                  | The network port number on which a server process listens for incoming connections from clients.                                         |
+| NODE_ENV     | Yes                                      | prod                                                  | Indicate the name of your .env file you want to use.                                                                                     |
+
+### Build for production
+
+Set up Environment Variables: Create a `prod.env` file in the root of the project directory to store your
```
Comment (k-y0u) [inline README.md:79]: remove `Set up Environment Variables:`

Comment (k-y0u) [inline README.md:79]: `prod.env` -> `[env_name].env`

File: README.md (hunk 23 above)
Comment (Zorin95670) [reply README.md:43]: done

File: README.md (hunk 24 above)
Comment (Zorin95670) [reply README.md:53]: done

File: README.md (hunk 25 above)
Comment (Zorin95670) [reply README.md:57]: done

File: README.md (hunk 26 above)
Comment (Zorin95670) [reply README.md:97]: done

File: README.md (hunk 27 above)
Comment (Zorin95670) [reply README.md:79]: done

Comment (Zorin95670) [reply README.md:79]: done

File: README.md [hunk 28]
```diff
@@ -1,2 +1,104 @@
...
+configuration settings. Add the following variables to the .env file:
+
+```makefile
+DATABASE_URI=postgres://leto_admin:password@localhost:5432/leto_db
+APP_ID=my_app_id
+MASTER_KEY=my_super_secret_master_key
+SERVER_URL=http://localhost:1337/api
+PARSE_MOUNT=/api
+PORT=1337
+NODE_ENV=prod
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply README.md:89]: done

File: src/index.js (hunk 2 above)
Comment (Zorin95670) [reply src/index.js:4]: so we can't use src because it need to be babelize

File: README.md (hunk 26 above)
Comment (k-y0u) [reply README.md:103]: nope

**Review by k-y0u — APPROVED** [review]

**Review by justinehell — APPROVED** [review]

**Review by AlexisNeveu — APPROVED** [review]

File: jest.config.cjs [hunk 29]
```diff
@@ -0,0 +1,31 @@
...
+  coveragePathIgnorePatterns: ['/node_modules/'],
+  testMatch: [
+    '<rootDir>/tests/unit/**/*.spec.js',
+  ],
+  moduleFileExtensions: ['js'],
+  moduleNameMapper: {
+    '^src/(.*)$': '<rootDir>/src/$1',
+    '^tests/(.*)$': '<rootDir>/tests/$1'
+  },
+  // transform: {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply jest.config.cjs:22]: done

File: src/models/Configuration.js [hunk 30]
```diff
@@ -0,0 +1,89 @@
...
+ * All api configuration keys and values.
+ */
+class Configuration {
+  /**
+   * Default constructor.
+   * @param {object} [props] - Object that contains all properties to set.
+   * @param {string} [props.mode] - The name of your .env file and running mode.
+   * @param {string} [props.mountPath] - That specifies the mount path where the Parse middleware
+   * should be deployed in your Node.js application.
+   * @param {number} [props.port] - The network port number on which a server process listens
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/Configuration.js:11]: it's good.

**Review by justinehell — APPROVED** [review]

**Review by aboubakar14 — APPROVED** [review]

**Pull request merged**